modules/ensemble/lib/widget/tooltip.dart

```diff
@@ -0,0 +1,182 @@
+import 'package:ensemble/framework/extensions.dart';
+import 'package:ensemble/widget/calendar.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:ensemble/framework/widget/widget.dart';
+import 'package:ensemble/framework/action.dart';
+import 'package:ensemble/screen_controller.dart';
+import 'package:ensemble/util/utils.dart';
+import 'package:ensemble_ts_interpreter/invokables/invokable.dart';
+import 'package:ensemble/framework/view/data_scope_widget.dart';
+import 'package:ensemble/page_model.dart';
+import 'package:ensemble/framework/widget/has_children.dart';
+import 'package:ensemble/framework/widget/view_util.dart';
+import 'package:ensemble/framework/device.dart';
+import 'package:yaml/yaml.dart';
+
+
+class ToolTip extends StatefulWidget
+    with Invokable, HasController<ToolTipController, ToolTipState> {
+  static const type = 'ToolTip';
+
+  ToolTip({Key? key}) : super(key: key);
+
+  final ToolTipController _controller = ToolTipController();
+
+  @override
+  ToolTipController get controller => _controller;
+
+  @override
+  State<StatefulWidget> createState() => ToolTipState();
+
+  @override
+  Map<String, Function> getters() {
+    return {
+      'isVisible': () => _controller.isVisible,
+    };
+  }
+
+  @override
+  Map<String, Function> setters() {
+    return {
+      'message': (value) => _controller.message = Utils.optionalString(value),
+      'child': (value) => _controller.child = value,
+      'textStyle': (value) => _controller.textStyle = Utils.getTextStyle(value),
+      'height': (value) => _controller.height = Utils.optionalDouble(value),
+      'padding': (value) => _controller.padding = Utils.getInsets(value),
+      'margin': (value) => _controller.margin = Utils.getInsets(value),
+      'verticalOffset': (value) => _controller.verticalOffset = Utils.optionalDouble(value),
+      'preferBelow': (value) => _controller.preferBelow = Utils.optionalBool(value),
+      'excludeFromSemantics': (value) => _controller.excludeFromSemantics = Utils.optionalBool(value),
+      'waitDuration': (value) => _controller.waitDuration = Utils.getDuration(value),
+      'showDuration': (value) => _controller.showDuration = Utils.getDuration(value),
+      'triggerMode': (value) => _controller.triggerMode = TooltipTriggerMode.values.from(value),
+      'onShow': (definition) => _controller.onShow = EnsembleAction.from(definition, initiator: this),
+      'onDismiss': (definition) => _controller.onDismiss = EnsembleAction.from(definition, initiator: this),
+      'decoration': (value) => _controller.setDecoration(value),
+    };
+  }
+
+  @override
+  Map<String, Function> methods() {
+    return {
+      'show': () => _controller.show(),
+      'hide': () => _controller.hide(),
+    };
+  }
+}
+
+class ToolTipController extends Controller {
```

Comment:
see if you can extend one of the existing controllers that already capture these properties

Reply:
Yeah, sorry for oversight, i implemented `BoxController` and handled box decoration and other properties from it.